db/view/view.cc

```diff
@@ -1772,15 +1774,36 @@ get_view_natural_endpoint(
         }
     }
 
-    SCYLLA_ASSERT(base_endpoints.size() == view_endpoints.size());
     auto base_it = std::find(base_endpoints.begin(), base_endpoints.end(), me);
     if (base_it == base_endpoints.end()) {
         // This node is not a base replica of this key, so we return empty
         // FIXME: This case shouldn't happen, and if it happens, a view update
         // would be lost.
         ++cf_stats.total_view_updates_on_wrong_node;
-        return {};
+        return std::make_pair(std::nullopt, std::nullopt);
+    }
+
+    // The following code handles the case where after a replication factor change
+    // the base replica completed the corresponding tablet migration at a different
+    // time than the view replica.
+    // RF can be changed only by 1 at a time, so the difference between the number
+    // of base and view replicas should be at most 1.
+    SCYLLA_ASSERT(std::abs((int)(base_endpoints.size() - view_endpoints.size())) <= 1);
+    std::optional<gms::inet_address> extra_view_endpoint;
+    if (base_endpoints.size() > view_endpoints.size()) {
+        // Because RF changes only affect the last replica in the list, we can simply
+        // ignore the extra base replica.
+        base_endpoints.pop_back();
+        ++cf_stats.total_base_view_replicas_mismatch;
```

Comment:
Is it safe to work with this assumption that last element is popped off?

what if we have RF=4, 3 racks

rack1 { n1 } rack2 { n20, n21 }, rack3 { n3}

and a given tablet has following replicas { n1, n20, n21, n3}

not sure it's the case today, but thinking about the future, it's perfectly possible we remove a replica from the rack with highest load for a given tablet.

we do have get_pending_endpoints(token) in effective_replication_map, later we may want to introduce get_leaving_endpoints(token), so we don't have to work with assumption here, but rather rely on the source of truth.

/cc @ptrsmrn

Reply:
I believe currently, if we remove one replica, we'll first remove the last replica from the list, and afterwards the replicas will get rebalanced between racks in another migration.

If we remove some replica other than the last one, the base/view pairing will change. Currently, if pairing changes during a migration, we can lose one view update update (I described it briefly in https://github.com/scylladb/scylladb/pull/21453#discussion_r1849320900, 2nd last paragraph).

Actually, perhaps taking leaving endpoints (or endpoints being replaced in general) into account could help in both cases - during a migration we would additionally send updates to view replicas originally paired with the replaced base replicas.

I'll try implementing this, as it could also help with base/view consistency issues in https://github.com/scylladb/scylladb/pull/21453

However, if we don't manage to reproduce this issue with decreasing RF, I think I'll send it as a separate patch

Reply:
> Actually, perhaps taking leaving endpoints (or endpoints being replaced in general) into account could help in both cases - during a migration we would additionally send updates to view replicas originally paired with the replaced base replicas.
> I'll try implementing this, as it could also help with base/view consistency issues in https://github.com/scylladb/scylladb/pull/21453

After a second thought, that unfortunately won't help for pairing changes in general - in these cases we may lose updates to view replicas completely unrelated to the migrated replicas.